### DISCUSSION TITLE
Evaluate using Profile-Guided Optimization (PGO) and Post-Link Optimization (PLO)

### DISCUSSION THREAD

**Discussion by zamazan4ik** [DataDog/datadog-static-analyzer#377] [Ideas]
Hi!

Recently I did a lot of PGO-related benchmarks on different software - all available results are [here](https://github.com/zamazan4ik/awesome-pgo). According to my tests, PGO can help with achieving better performance in many workloads, including compiler-like (compilers, static analysis, code formatters, etc.). Since this, I think it would be a good idea to test PGO with the analyzer to potentially improve its performance.

I can suggest to do the following things:

* Evaluate PGO on the analyzer
* If it benefits the project - add a note to the documentation about building with PGO. In this case, users and maintainers who build their own packages (if any) will be aware of PGO as an additional way to optimize the analyzer
* Optimize provided by Datadog team binaries on the CI (like it's already done for other projects like Rustc)
* After that, I suggest trying to apply [LLVM BOLT](https://github.com/llvm/llvm-project/blob/main/bolt/README.md) as an additional post-PGO step. Rustc already does it on some platforms

For the Rust projects, I suggest PGO optimizing with [cargo-pgo](https://github.com/Kobzol/cargo-pgo).

**Comment by jasonforal** [comment]
Interesting -- thanks for the share.

It looks like we'd run into a similar issue as https://github.com/Kobzol/cargo-pgo/issues/38, as a majority of our CPU time is spent using the `tree-sitter` library.

But this is a nice idea to keep on the radar
